# Machine Learning - Final project
### Empathy Prediction using the [Young people survey](https://www.kaggle.com/miroslavsabo/young-people-survey/) dataset

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

sns.set(style="white", color_codes=True)
sns.set_context(rc={"font.family":'sans',"font.size":24,"axes.titlesize":24,"axes.labelsize":24})  
pd.options.display.max_columns = 150
pd.options.display.max_rows = 150

#### Loading datasets

In [89]:
df = pd.read_csv('./input/responses.csv')
columns_desc = pd.read_csv('./input/columns.csv')

In [90]:
df.shape

(1010, 150)

#### Defining way to access description easily

In [70]:
all_columns =list(df)

desc = {}
for col_name in all_columns:
    desc[col_name] = columns_desc.loc[columns_desc['short'] == col_name].iloc[0][0]
    
def print_desc(column_name):
    print(column_name+" --> "+desc[column_name])

Example:

In [71]:
for col in all_columns[:5]:
    print_desc(col)

Music --> I enjoy listening to music.
Slow songs or fast songs --> I prefer.
Dance --> Dance, Disco, Funk
Folk --> Folk music
Country --> Country


## Data Exploration and preprocessing

In [143]:
df.describe()

Music  Slow songs or fast songs        Dance         Folk  \
count  1005.000000               1005.000000  1005.000000  1005.000000   
mean      4.736318                  3.330348     3.112438     2.288557   
std       0.659835                  0.831600     1.170817     1.137166   
min       1.000000                  1.000000     1.000000     1.000000   
25%       5.000000                  3.000000     2.000000     1.000000   
50%       5.000000                  3.000000     3.000000     2.000000   
75%       5.000000                  4.000000     4.000000     3.000000   
max       5.000000                  5.000000     5.000000     5.000000   

           Country  Classical music      Musical          Pop         Rock  \
count  1005.000000      1005.000000  1005.000000  1005.000000  1005.000000   
mean      2.123383         2.956219     2.763184     3.470647     3.773134   
std       1.075210         1.244939     1.261457     1.159743     1.183500   
min       1.000000         1.000000     1.000000     1.000000     1.000000   
25%       1.000000         2.000000     2.000000     3.000000     3.000000   
50%       2.000000         3.000000     3.000000     4.000000     4.000000   
75%       3.000000         4.000000     4.000000     4.000000     5.000000   
max       5.000000         5.000000     5.000000     5.000000     5.000000   

       Metal or Hardrock         Punk  Hiphop, Rap  Reggae, Ska  Swing, Jazz  \
count        1005.000000  1005.000000  1005.000000  1005.000000  1005.000000   
mean            2.358209     2.446766     2.915423     2.770149     2.757214   
std             1.375310     1.303688     1.374139     1.210802     1.253187   
min             1.000000     1.000000     1.000000     1.000000     1.000000   
25%             1.000000     1.000000     2.000000     2.000000     2.000000   
50%             2.000000     2.000000     3.000000     3.000000     3.000000   
75%             3.000000     3.000000     4.000000     4.000000     4.000000   
max             5.000000     5.000000     5.000000     5.000000     5.000000   

       Rock n roll  Alternative       Latino  Techno, Trance        Opera  \
count  1005.000000  1005.000000  1005.000000     1005.000000  1005.000000   
mean      3.140299     2.830846     2.834826        2.333333     2.141294   
std       1.231726     1.341657     1.324223        1.325258     1.185880   
min       1.000000     1.000000     1.000000        1.000000     1.000000   
25%       2.000000     2.000000     2.000000        1.000000     1.000000   
50%       3.000000     3.000000     3.000000        2.000000     2.000000   
75%       4.000000     4.000000     4.000000        3.000000     3.000000   
max       5.000000     5.000000     5.000000        5.000000     5.000000   

            Movies       Horror     Thriller       Comedy     Romantic  \
count  1005.000000  1005.000000  1005.000000  1005.000000  1005.000000   
mean      4.615920     2.793035     3.384080     4.497512     3.488557   
std       0.694131     1.412435     1.197501     0.777705     1.209655   
min       1.000000     1.000000     1.000000     1.000000     1.000000   
25%       4.000000     1.000000     3.000000     4.000000     3.000000   
50%       5.000000     3.000000     4.000000     5.000000     4.000000   
75%       5.000000     4.000000     4.000000     5.000000     5.000000   
max       5.000000     5.000000     5.000000     5.000000     5.000000   

            Sci-fi          War  Fantasy/Fairy tales     Animated  \
count  1005.000000  1005.000000          1005.000000  1005.000000   
mean      3.112438     3.153234             3.751244     3.790050   
std       1.310506     1.346532             1.182463     1.220958   
min       1.000000     1.000000             1.000000     1.000000   
25%       2.000000     2.000000             3.000000     3.000000   
50%       3.000000     3.000000             4.000000     4.000000   
75%       4.000000     4.000000             5.000000     5.000000   
max      

In [161]:
df.rename(columns={'Left - right handed': 'LeftRightHanded', 'Village - town': 'VillageTown', 'Internet usage': 'InternetUsage', 'Only child': 'OnlyChild', 'House - block of flats': 'HouseFlats'}, inplace=True)

I notice that some attributes like Music have really skewed distributions, later on I could try to apply some log transformations and compare the performances

### Missing values

Let's start by looking at how many rows are missing a class label, I will delete these rows since it makes no sense to impute them, train or test using them

In [86]:
df['Empathy'].isna().sum()

5

In [91]:
df.dropna(subset=['Empathy'], inplace=True)

In [92]:
df.shape

(1005, 150)

Let's now look at the percentage of missing values in all columns in desc order

In [94]:
df_na = (df.isnull().sum() / len(df)) * 100
df_na = df_na.drop(df_na[df_na == 0].index).sort_values(ascending=False)
missing_data = pd.DataFrame({'Missing Ratio': df_na})
missing_data.head(150).T

Height   Weight  Passive sport  Chemistry  Geography  Theatre  \
Missing Ratio  1.99005  1.99005       1.492537   0.995025   0.895522  0.79602   

               Documentary  Smoking   Latino     Punk  Criminal damage  \
Missing Ratio      0.79602  0.79602  0.79602  0.79602         0.696517   

               Compassion to animals  Final judgement  Reggae, Ska  Gardening  \
Missing Ratio               0.696517         0.696517     0.696517   0.696517   

               Alternative  Rock n roll  Techno, Trance       Age  \
Missing Ratio     0.696517     0.696517        0.696517  0.696517   

               Classical music  Swing, Jazz    Movies   Biology   Reading  \
Missing Ratio         0.696517     0.597015  0.597015  0.597015  0.597015   

                 Giving        PC    Gender  Number of siblings  \
Missing Ratio  0.597015  0.597015  0.597015            0.597015   

               Responding to a serious letter  Daily events  \
Missing Ratio                        0.597015      0.597015   

               Science and technology  Art exhibitions  Friends versus money  \
Missing Ratio                0.597015         0.597015              0.597015   

                Writing  Countryside, outdoors      Rock      Folk  \
Missing Ratio  0.597015               0.597015  0.597015  0.497512   

               Self-criticism   Spiders   Country  Economy Management  \
Missing Ratio        0.497512  0.497512  0.497512            0.497512   

               Energy levels  Workaholism  Psychology  Foreign languages  \
Missing Ratio       0.497512     0.497512    0.497512           0.497512   

               Prioritising workload  Getting up  Medicine   Alcohol  \
Missing Ratio               0.497512    0.497512  0.497512  0.497512   

               Socializing  Village - town     Pets  Active sport  \
Missing Ratio     0.497512         0.39801  0.39801       0.39801   

               Reliability  Loss of interest  Fun with friends     Cars  \
Missing Ratio      0.39801           0.39801           0.39801  0.39801   

               House - block of flats  Funniness  Decision making    Dance  \
Missing Ratio                 0.39801    0.39801          0.39801  0.39801   

               Hiphop, Rap  Questionnaires or polls  Finding lost valuables  \
Missing Ratio      0.39801                  0.39801                 0.39801   

               Personality  Small - big dogs  Happiness in life  \
Missing Ratio      0.39801           0.39801            0.39801   

               Getting angry  Children  Mood swings  Western  Internet  \
Missing Ratio        0.39801   0.39801      0.39801  0.39801   0.39801   

               Cheating in school  Hypochondria  Judgment calls  Romantic  \
Missing Ratio             0.39801       0.39801         0.39801  0.298507   

                 Comedy  Fantasy/Fairy tales  Animated  Religion   Physics  \
Missing Ratio  0.298507             0.298507  0.298507  0.298507  0.298507   

                Dancing  Metal or Hardrock       Pop  Mathematics     Music  \
Missing Ratio  0.298507           0.298507  0.298507     0.298507  0.298507   

               Appearence and gestures  Spending on looks  Life struggles  \
Missing Ratio                 0.298507           0.298507        0.298507   

               Interests or hobbies   Charity  Finances  Elections  \
Missing Ratio              0.298507  0.298507  0.298507   0.298507   

               Thinking ahead  Writing notes  Healthy eating  \
Missing Ratio        0.298507       0.298507        0.298507   

               Entertainment spending      Rats   Heights  Left - right handed  
Missing Ratio                0.298507  0.298507  0.298507             0.298507

The percentage of missing values is not a lot, we can easily impute them without introducing too much bias by using the mode

In [95]:
df[all_columns]=df[all_columns].fillna(df.mode().iloc[0])

In [97]:
df.isnull().any().any()

False

There are no missing values now

### Data types

In [17]:
df['Music'] = df['Music'].astype(int)

['a', 'b', 'c', 'd']

In [105]:
df.dtypes[:10]

Music                       float64
Slow songs or fast songs    float64
Dance                       float64
Folk                        float64
Country                     float64
Classical music             float64
Musical                     float64
Pop                         float64
Rock                        float64
Metal or Hardrock           float64
dtype: object

In [109]:
float_cols = df.dtypes == np.float64

In [115]:
fake_float = list(df.loc[:, float_cols])

In [116]:
len(fake_float)

134

134 columns are float but should be int since the decimal value is always 0, let's convert them

In [117]:
df[fake_float] = df[fake_float].astype(int)

### One hot encoding
Since the algorithms that I have in mind do not handle categorical values I will need to perform one hot encoding or encode values in numbers for ordinal data

Let's look at categorical columns and decide which ones can be ordinal

In [163]:
categorical = df.dtypes == np.object

In [164]:
df.loc[:, categorical].T

0                        1     \
Smoking                     never smoked             never smoked   
Alcohol                      drink a lot              drink a lot   
Punctuality          i am always on time         i am often early   
Lying                              never                sometimes   
InternetUsage            few hours a day          few hours a day   
Gender                            female                   female   
LeftRightHanded             right handed             right handed   
Education        college/bachelor degree  college/bachelor degree   
OnlyChild                             no                       no   
VillageTown                      village                     city   
HouseFlats                block of flats           block of flats   

                                    2                              3     \
Smoking                    tried smoking                  former smoker   
Alcohol                      drink a lot                    drink a lot   
Punctuality      i am often running late               i am often early   
Lying                          sometimes  only to avoid hurting someone   
InternetUsage            few hours a day                most of the day   
Gender                            female                         female   
LeftRightHanded             right handed                   right handed   
Education               secondary school        college/bachelor degree   
OnlyChild                             no                            yes   
VillageTown                         city                           city   
HouseFlats                block of flats                 house/bungalow   

                                  4                              5     \
Smoking                  tried smoking                   never smoked   
Alcohol                 social drinker                          never   
Punctuality        i am always on time               i am often early   
Lying            everytime it suits me  only to avoid hurting someone   
InternetUsage          few hours a day                few hours a day   
Gender                          female                           male   
LeftRightHanded           right handed                   right handed   
Education             secondary school               secondary school   
OnlyChild                           no                             no   
VillageTown                    village                           city   
HouseFlats              house/bungalow                 block of flats   

                                    6                        7     \
Smoking                    tried smoking           current smoker   
Alcohol                   social drinker              drink a lot   
Punctuality             i am often early      i am always on time   
Lying                              never                sometimes   
InternetUsage    less than an hour a day          few hours a day   
Gender                            female                     male   
LeftRightHanded             right handed             right handed   
Education               secondary school  college/bachelor degree   
OnlyChild                             no                       no   
VillageTown                      village                     city   
HouseFlats                house/bungalow           house/bungalow   

                             8                        9     \
Smoking             tried smoking             never smoked   
Alcohol            social drinker              drink a lot   
Punctuality      i am often early  i am often running late   
Lying                   sometimes                sometimes   
InternetUsage     few hours a day          few hours a day   
Gender                     female                   female   
LeftRightHanded      right handed             right handed   
Education        secondary school         secondary school   
OnlyChild                      no         

In [128]:
categorical_cols = list(df.loc[:, categorical])

In [215]:
for col in categorical_cols:
    print_desc(col)

Smoking --> Smoking habits
Alcohol --> Drinking
Punctuality --> Timekeeping.
Lying --> Do you lie to others?
Internet usage --> How much time do you spend online?
Gender --> Gender
Left - right handed --> I am
Education --> Highest education achieved
Only child --> I am the only child
Village - town --> I spent most of my childhood in a
House - block of flats --> I lived most of my childhood in a


There are a lot of ordinal columns, which is better because we will not occur in a feature explosion using OHE, I think the ordinal values could be:

* Punctuality
* InternetUsage
* Education

Some other that only have True/False or Yes/No or in general 2 possible values can be encoded using the same attribute and an integer value (0 or 1), these features are:

* Gender
* LeftRightHanded
* OnlyChild
* VillageTown
* HouseFlats

The remaining categorical attributes are the ones that I think are the most important to predict empathy or that are not really ordinal and I would not want to introduce a bias by considering them as ordinal, these attributes are:

* Smoking
* Lying
* Alcohol

In [145]:
print("Unique values of Smoking")
print(df.Smoking.unique())

print("\nUnique values of Lying")
print(df.Lying.unique())


Unique values of Smoking
['never smoked' 'tried smoking' 'former smoker' 'current smoker']

Unique values of Lying
['never' 'sometimes' 'only to avoid hurting someone'
 'everytime it suits me']


In [153]:
df.shape

(1005, 158)

In [148]:
def OHE_Smoking_N(value):
    return 1 if value=='never smoked' else 0

def OHE_Smoking_T(value):
    return 1 if value=='tried smoking' else 0

def OHE_Smoking_F(value):
    return 1 if value=='former smoker' else 0

def OHE_Smoking_C(value):
    return 1 if value=='current smoker' else 0

In [149]:
df['Smoking_Never'] = df.Smoking.apply(OHE_Smoking_N)
df['Smoking_Tried'] = df.Smoking.apply(OHE_Smoking_T)
df['Smoking_Former'] = df.Smoking.apply(OHE_Smoking_F)
df['Smoking_Current'] = df.Smoking.apply(OHE_Smoking_C)

In [151]:
def OHE_Lying_N(value):
    return 1 if value=='never' else 0

def OHE_Lying_S(value):
    return 1 if value=='sometimes' else 0

def OHE_Lying_O(value):
    return 1 if value=='only to avoid hurting someone' else 0

def OHE_Lying_E(value):
    return 1 if value=='everytime it suits me' else 0

In [152]:
df['Lying_Never'] = df.Lying.apply(OHE_Lying_N)
df['Lying_Sometimes'] = df.Lying.apply(OHE_Lying_S)
df['Lying_Only'] = df.Lying.apply(OHE_Lying_O)
df['Lying_Everytime'] = df.Lying.apply(OHE_Lying_E)

In [188]:
def OHE_Alcohol_D(value):
    return 1 if value=='drink a lot' else 0

def OHE_Alcohol_S(value):
    return 1 if value=='social drinker' else 0

def OHE_Alcohol_N(value):
    return 1 if value=='never' else 0

In [189]:
df['Alcohol_Drink'] = df.Alcohol.apply(OHE_Alcohol_D)
df['Alcohol_Social'] = df.Alcohol.apply(OHE_Alcohol_S)
df['Alcohol_Never'] = df.Alcohol.apply(OHE_Alcohol_N)

#### True False OHE

In [171]:
def OHE_Male(value):
    return 1 if value=='male' else 0

In [172]:
df['Male'] = df.Gender.apply(OHE_Male)

In [166]:
def OHE_Right(value):
    return 1 if value=='right handed' else 0

In [167]:
df['Right_Handed'] = df.LeftRightHanded.apply(OHE_Right)

In [174]:
def OHE_Only(value):
    return 1 if value=='yes' else 0

In [175]:
df['Only_Child'] = df.OnlyChild.apply(OHE_Only)

In [182]:
def OHE_City(value):
    return 1 if value=='city' else 0

In [183]:
df['City'] = df.VillageTown.apply(OHE_City)

In [184]:
def OHE_House(value):
    return 1 if value=='house/bungalow' else 0

In [185]:
df['House'] = df.HouseFlats.apply(OHE_House)

#### Ordinal

In [192]:
print("Unique values of Punctuality")
print(df.Punctuality.unique())

Unique values of Punctuality
['i am always on time' 'i am often early' 'i am often running late']


In [194]:
def OHE_Punctuality_Ordinal(value):
    if value=='i am always on time':
        return 1
    if value=='i am often early':
        return 2 
    if value=='i am often running late':
        return 0
    #should not arrive here
    return 0

In [195]:
df['Punctual'] = df.Punctuality.apply(OHE_Punctuality_Ordinal)

In [197]:
print("Unique values of InternetUsage")
print(df.InternetUsage.unique())

Unique values of InternetUsage
['few hours a day' 'most of the day' 'less than an hour a day'
 'no time at all']


In [198]:
def OHE_InternetUsage_Ordinal(value):
    if value=='most of the day':
        return 3
    if value=='few hours a day':
        return 2 
    if value=='less than an hour a day':
        return 1
    if value=='no time at all':
        return 0
    #should not arrive here
    return 0

In [199]:
df['InternetUsage'] = df.InternetUsage.apply(OHE_InternetUsage_Ordinal)

In [202]:
print("Unique values of Education")
print(df.Education.unique())

Unique values of Education
['college/bachelor degree' 'secondary school' 'primary school'
 'masters degree' 'doctorate degree' 'currently a primary school pupil']


In [203]:
def OHE_Education_Ordinal(value):
    if value=='doctorate degree':
        return 5
    if value=='masters degree':
        return 4 
    if value=='college/bachelor degree':
        return 3
    if value=='secondary school':
        return 2
    if value=='primary school':
        return 1
    if value=='currently a primary school pupil':
        return 0
    #should not arrive here
    return 0

In [204]:
df['Education'] = df.Education.apply(OHE_Education_Ordinal)

#### Deleting categorical columns that were encoded

In [208]:
encoded_columns = ['Smoking','Lying','Alcohol','Gender','LeftRightHanded','OnlyChild','VillageTown','HouseFlats','Punctuality']

In [211]:
df = df.drop(encoded_columns, axis=1)

In [213]:
df.shape

(1005, 158)

In [216]:
pd.options.display.max_columns = 158
pd.options.display.max_rows = 158

In [218]:
df.head()

Music  Slow songs or fast songs  Dance  Folk  Country  Classical music  \
0      5                         3      2     1        2                2   
1      4                         4      2     1        1                1   
2      5                         5      2     2        3                4   
3      5                         3      2     1        1                1   
4      5                         3      4     3        2                4   

   Musical  Pop  Rock  Metal or Hardrock  Punk  Hiphop, Rap  Reggae, Ska  \
0        1    5     5                  1     1            1            1   
1        2    3     5                  4     4            1            3   
2        5    3     5                  3     4            1            4   
3        1    2     2                  1     4            2            2   
4        3    5     3                  1     2            5            3   

   Swing, Jazz  Rock n roll  Alternative  Latino  Techno, Trance  Opera  \
0            1            3            1       1               1      1   
1            1            4            4       2               1      1   
2            3            5            5       5               1      3   
3            1            2            5       1               2      1   
4            2            1            2       4               2      2   

   Movies  Horror  Thriller  Comedy  Romantic  Sci-fi  War  \
0       5       4         2       5         4       4    1   
1       5       2         2       4         3       4    1   
2       5       3         4       4         2       4    2   
3       5       4         4       3         3       4    3   
4       5       4         4       5         2       3    3   

   Fantasy/Fairy tales  Animated  Documentary  Western  Action  History  \
0                    5         5            3        1       2        1   
1                    3         5            4        1       4        1   
2                    5         5            2        2       1        1   
3                    1         2            5        1       2        4   
4                    4         4            3        1       4        3   

   Psychology  Politics  Mathematics  Physics  Internet  PC  \
0           5         1            3        3         5   3   
1           3         4            5        2         4   4   
2           2         1            5        2         4   2   
3           4         5            4        1         3   1   
4           2         3            2        2         2   2   

   Economy Management  Biology  Chemistry  Reading  Geography  \
0                   5        3          3        3          3   
1                   5        1          1        4          4   
2                   4        1          1        5          2   
3                   2        3          3        5          4   
4                   2        3          3        5          2   

   Foreign languages  Medicine  Law  Cars  Art exhibitions  Religion  \
0                  5         3    1     1                1         1   
1                  5         1    2     2                2         1   
2                  5         2    3     1                5         5   
3                  4         2    5     1                5         4   
4                  3         3    2     3                1         4   

   Countryside, outdoors  Dancing  Musical instruments  Writing  \
0                      5        3                    3        2   
1                      1        1                    1        1   
2                      5        5                    5        5   
3                      1        1                    1        3   
4                      4        1                    3        1   

   Passive sport  Active sport  Gardening  Celebrities  Shopping  \
0              1             5          5            1         4   
1              1             1          1            2         3  